op-chain-ops: decouple from bindings

tynes opened the pull request:
**Description**

Another commit that removes bindings from `op-chain-ops`.
The entire repo is very close to being completely decoupled
from having a global set of bindings. This does have a little
bit of code duplication, but it generally results in a more
simple layout.

Both the `crossdomain` and `upgrades` package end up with their
own `bindings` package. The `crossdomain` package is mostly
legacy but the `indexer` still depends on it. The `indexer`
is in maintenance mode, after reth exex it makes a lot of sense
to build indexers directly into reth. The `crossdomain` package
uses the bindings to migrate and serialize L1 to L2 cross domain
messages. These bindings never need to be updated.

The `upgrades` package has a justfile added that can be used to
regenerate the bindings. This makes it easier to generate bindings
based on a specific release. Previously the `upgrades` bindings
were coupled in a strange way and backwards compatibility
hacks needed to exist. Now the team working on upgrades can update
the bindings however they want so the proper upgrade can easily
be generated.

<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

tynes committed:
op-chain-ops: decouple from bindings

Another commit that removes bindings from `op-chain-ops`.
The entire repo is very close to being completely decoupled
from having a global set of bindings. This does have a little
bit of code duplication, but it generally results in a more
simple layout.

Both the `crossdomain` and `upgrades` package end up with their
own `bindings` package. The `crossdomain` package is mostly
legacy but the `indexer` still depends on it. The `indexer`
is in maintenance mode, after reth exex it makes a lot of sense
to build indexers directly into reth. The `crossdomain` package
uses the bindings to migrate and serialize L1 to L2 cross domain
messages. These bindings never need to be updated.

The `upgrades` package has a justfile added that can be used to
regenerate the bindings. This makes it easier to generate bindings
based on a specific release. Previously the `upgrades` bindings
were coupled in a strange way and backwards compatibility
hacks needed to exist. Now the team working on upgrades can update
the bindings however they want so the proper upgrade can easily
be generated.

mds1 commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -757,9 +757,9 @@ func SystemConfig(batch *safe.Batch, implementations superchain.ImplementationLi
 			L1CrossDomainMessenger:       common.Address(list.L1CrossDomainMessengerProxy),
 			L1ERC721Bridge:               common.Address(list.L1ERC721BridgeProxy),
 			L1StandardBridge:             common.Address(list.L1StandardBridgeProxy),
-			L2OutputOracle:               common.Address(list.L2OutputOracleProxy),
```
I think the DGF should be in here: https://github.com/ethereum-optimism/optimism/blob/f2e06c31b8cac95d36f2219a468244163f8bc26f/packages/contracts-bedrock/src/L1/SystemConfig.sol#L38

tynes replied:
Has the dgf been added to the superchain registry yet? If no, then I would like to save that for the future

mds1 replied:
It has been added for Sepolia:
- https://github.com/ethereum-optimism/superchain-registry/blob/110e744c97b4873384ad2da365c281639fc0668e/superchain/extra/addresses/sepolia/op.json#L12-L14
- https://github.com/ethereum-optimism/superchain-registry/blob/110e744c97b4873384ad2da365c281639fc0668e/superchain/implementations/networks/sepolia.yaml#L15-L16

tynes replied:
In [09f974b](https://github.com/ethereum-optimism/optimism/pull/10466/commits/09f974b31d3b1b144c3605aa1ff9e0ad0b5c904f) i have added an always error case until the dispute game factory address can be accessed from the `ContractImplementations`, see https://github.com/ethereum-optimism/superchain-registry/blob/110e744c97b4873384ad2da365c281639fc0668e/superchain/superchain.go#L253

tynes replied:
Opened https://github.com/ethereum-optimism/superchain-registry/pull/227 to address this

tynes committed:
upgrades: error case

Since the AddressList doesn't have the `DisputeGameFactory`
yet, it cannot be pulled into the upgrade scripts. Instead
return an error so that we MUST update it.

https://github.com/ethereum-optimism/superchain-registry/blob/110e744c97b4873384ad2da365c281639fc0668e/superchain/superchain.go#L194

tynes commented:
@mds1 I'd like to merge this as is, the superchain ops repo isn't ready for the change you suggested
